---
title:  "New MegaMUGA/GigaMUGA annotations"
author: "Karl Broman"
date:   "`r Sys.Date()`"
output:
    html_document:
        code_folding: hide
        toc: true
        toc_float: true
        toc_depth: 3
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.width=11, fig.height=6.5,
                      message=FALSE, warning=FALSE)
options(width=110)
```


## Introduction

While preparing a
[talk](https://kbroman.org/Talk_CTC2018/broman_ctc2018.pdf) and
[paper](https://www.biostat.wisc.edu/~kbroman/publications/mpp_diag.pdf)
on genotype diagnostics in Diversity Outbred mice,
I developed some concerns about the UNC annotation file for the
GigaMUGA array. In particular, I felt that some of the columns may
have been scrambled:

- In blasting the GigaMUGA probe sequences against
  the mouse genome, I found the same number of non-unique markers
  (3,114), but they were not the same markers.

- Of the markers in common between the MegaMUGA and GigaMUGA arrays,
  the chromosome assignments have a lot of differences.

In case that the probe sequences in the UNC annotation files might be
in error, I obtained the array probe sequences directly from GeneSeek.
These proved to be the same as those in the UNC files.

My goal here is to construct new annotation files for the
GigaMUGA and MegaMUGA arrays; with consistency between the two; with
updated information about which markers map uniquely, which map to
multiple locations, and which do not map; and including genomic
coordinates only for the markers that map uniquely.

A summary of my findings:

- For markers mapping uniquely, the genomic coordinates in the UNC
  GigaMUGA annotation file were correct.

- The UNC MegaMUGA annotation file had coordinates in a previous mouse
  genome build (mm9). For markers mapping uniquely, there are a small
  number with differences in chromosome assignments; there are lots of
  differences in positions, with the new build (mm10).

- The "unique" column in the GigaMUGA annotation file is messed up.

- On both arrays, there are a number of pairs of markers with common probe
  sequences. Many of these are intended duplicates, but some are not
  and have inconsistent information

- Of the markers in common between the two arrays, some have different
  probe sequences on the two arrays, and there are cases where the
  sequence on one array maps uniquely but the other does not.

- I'm not sure what to do with the markers on the "P" chromosome,
  which I understand to be the pseudoautosomal region, as the markers
  are not mapping to the mm10 mouse genome build.



## Procedures

The key thing here was to run blastn to map all array probe sequences
to the mouse genome (build mm10). There is some trickiness in
trimming SNP alleles off some problems, and in identifying the SNP
position from the blast results.


## Preliminaries

I'll first load some R packages.

```{r load_packages}
library(data.table)
library(broman)
library(devtools)
```

I'll now load various data sets: the GeneSeek and UNC annotation
files, and the blast results.

```{r load_data}
# GeneSeek files
gm_gs <- data.table::fread("../GeneSeek/gigamuga_geneseek.csv",
                           skip=7, data.table=FALSE)
wh <- which(gm_gs[,1]=="[Controls]")
gm_gs <- gm_gs[1:(wh-1),]
rownames(gm_gs) <- gm_gs$Name

mm_gs <- data.table::fread("../GeneSeek/megamuga_geneseek.csv",
                           skip=7, data.table=FALSE)
wh <- which(mm_gs[,1]=="[Controls]")
mm_gs <- mm_gs[1:(wh-1),]
rownames(mm_gs) <- mm_gs$Name

# UNC files
load("../UNC/snps.gigamuga.Rdata")
gm_unc <- snps
# omit "chr" from the chromosome IDs
gm_unc$chr <- sub("chr", "", gm_unc$chr)

load("../UNC/snps.megamuga.Rdata")
mm_unc <- snps
# omit "chr" from the chromosome IDs
mm_unc$chr <- sub("chr", "", mm_unc$chr)

# Blast results; keep just the perfect matches(?)
mm_blast <- readRDS("../Blast/results_mm/mm_blastn_results.rds")
mm_blast <- mm_blast[mm_blast$tot_mismatch==0,]
gm_blast <- readRDS("../Blast/results_gm/gm_blastn_results.rds")
gm_blast <- gm_blast[gm_blast$tot_mismatch==0,]
```


## Study blast results

```{r summarize_blast_results}
# no. blast hits
gm_tab <- table(gm_blast$query)
gm_nchr <- gm_nhits <- setNames(rep(0, nrow(gm_gs)), rownames(gm_gs))
gm_nhits[names(gm_tab)] <- gm_tab

# no. chromosomes hit
gm_tab_chr <- table(gm_blast$query, gm_blast$chr)
gm_nchr[rownames(gm_tab_chr)] <- rowSums(gm_tab_chr > 0)

# multiple hits, but all to the same chromosome
gm_mult_but1chr <- names(gm_nhits)[gm_nhits > 1 & gm_nchr==1]
ra <- sapply(gm_mult_but1chr, function(mar)
    diff(range(gm_blast$snp_pos[gm_blast$query==mar])))
# UNC19597786: two hits at exactly the same positions; why would blast do that?

# hit X and Y but no other chr
gm_mar_xy <- rownames(gm_tab_chr)[apply(gm_tab_chr, 1, function(a) all(a[1:20]==0) && all(a[21:22] > 0))]
# a bunch of these are named "PAR" but on GM they're mostly Y and a little X
# they mostly have multiple hits on X and Y
# in geneseek file, they're all on chr X
#
# 36 of GM markers with NA for chr have hits



# for MM, there are just 4 markers on chr P that are in the blast results
#  ... and all 2 map uniquely to chr 13; the other two map uniquely to chr 5.
# NNT001, NNT002 but while these are chr P in MM, they're chr 13 in GM
# SAbGeoEUCOMM001 and SAbGeoEUCOMM002 are chr P in MM, map uniquely to
# chr 5, and former is on chr 5 in GM while the latter is not present on GM
#
# only 4 of the P markers have hits
sum(mm_nhits[mm_unc$marker[mm_unc$chr=="P"]] > 0)

## Same for MegaMUGA
# no. blast hits
mm_tab <- table(mm_blast$query)
mm_nchr <- mm_nhits <- setNames(rep(0, nrow(mm_gs)), rownames(mm_gs))
mm_nhits[names(mm_tab)] <- mm_tab

# no. chromosomes hit
mm_tab_chr <- table(mm_blast$query, mm_blast$chr)
mm_nchr[rownames(mm_tab_chr)] <- rowSums(mm_tab_chr > 0)
```

```{r snp_pos}
gm_blast_uniq <- gm_blast[gm_blast$query %in% names(gm_nhits)[gm_nhits==1],]
rownames(gm_blast_uniq) <- gm_blast_uniq$query

gm_blast_uniq_chr <- setNames(gm_blast_uniq$chr, rownames(gm_blast_uniq))
gm_unc_uniq_chr <- gm_unc[rownames(gm_blast_uniq), "chr"]
gm_blast_uniq_pos <- setNames(gm_blast_uniq$snp_pos, rownames(gm_blast_uniq))
gm_unc_uniq_pos <- gm_unc[rownames(gm_blast_uniq), "pos"]


table(gm_blast_uniq_pos - gm_unc_uniq_pos, gm_gs[gm_blast_uniq$marker,"AlleleB_ProbeSeq"]=="")

# strands have no relationship :(
gm_blast_uniq_strand <- setNames(gm_blast_uniq$strand, rownames(gm_blast_uniq))
table(gm_gs[rownames(gm_blast_uniq),"IlmnStrand"], gm_blast_uniq_strand)


# the markers that are off by +/- 1 are exactly the ones with the AlleleB_ProbeSeq
# (and I think I'm right)
```

```{r p_mar}
# markers on p chr in MM that are also in GM
pmar_mm <- mm_unc$marker[mm_unc$chr=="P" & mm_unc$marker %in% gm_unc$marker]

# markers on p chr(?) ...well actually chr is NA... in GM that are also in MM
pmar_gm <- gm_unc$marker[is.na(gm_unc$chr) & gm_unc$marker %in% mm_unc$marker]

pmar_gm %wnin% pmar_mm # UNC28724761 UNC28725371 UNC28726152
# latter are on 18 in MM
pmar_mm %wnin% pmar_gm # NNT001      NNT002      SAbGeoEUCOMM001
# latter are on 5,13 in GM
```

```{r markers_with_common_probes}
# MM probes with the same sequence
tab <- table(mm_unc$seq.A)
dup <- names(tab)[tab > 1] # 51 sets
z <- lapply(dup, function(d) { z <- mm_unc[mm_unc$seq.A == d,]
         if(length(unique(z$chr[!is.na(z$chr)])) > 1) return(rownames(z)) })
z <- z[sapply(z, length)>0]

# MM UNC88694 (chr 1), UNC2512583 (chr 2)
[only 2nd one in GM]
# MM UNC23833145 (chr 14), UNC4368900 (chr 2)
[in GM, both said to be unique, but same probe and on different chromosomes]
# MM UNC20251282 (chr 11), UNC5746160 (chr 3)
[in GM, both said to be unique, but same probe and on different chromosomes]
# MM UNC27690624 (chr 17), UNC30009421 (chr 19)
[only 2nd one in GM]
# MM UNC19598208 (chr 11)  UNC24941679 (chr 14)
[in GM, both said to be unique, but same probe and on different chromosomes]

# GM has *a lot* of markers with the same sequence
tab <- table(gm_unc$seq.A)
dup <- names(tab)[tab > 1] # 2672 sets
z <- lapply(dup, function(d) { z <- gm_unc[gm_unc$seq.A == d,]
         if(length(unique(z$chr[!is.na(z$chr)])) > 1) return(rownames(z)) })
z <- z[sapply(z, length)>0]
# only 10 sets that are on different chromosomes: 3 above, and these 7:
# UNCrs52300311 complement008 complement018 complement016 (repetitive sequence that doesn't map)
# Mit016 Mit030             # Mit016 placed on M
# sanger1454  sanger1454q   # both map multiply, but in unc annotations said unique
# UNC9964748  UNC10237349   # "
# sanger2519  sanger2519q
# sanger2310  sanger2310q
# JAX00187253q JAX00187253
```




## Session info

Here are the versions of R and R packages that I am using.

```{r session_info}
devtools::session_info()
```
