---
title:  "New MegaMUGA/GigaMUGA annotations"
author: "Karl Broman"
date:   "`r Sys.Date()`"
output:
    html_document:
        code_folding: hide
        toc: true
        toc_float: true
        toc_depth: 3
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.width=11, fig.height=6.5,
                      message=FALSE, warning=FALSE)
options(width=110)
```


## Introduction

While preparing a
[talk](https://kbroman.org/Talk_CTC2018/broman_ctc2018.pdf) and
[paper](https://www.biostat.wisc.edu/~kbroman/publications/mpp_diag.pdf)
on genotype diagnostics in Diversity Outbred mice,
I developed some concerns about the UNC annotation file for the
GigaMUGA array. In particular, I felt that some of the columns may
have been scrambled:

- In blasting the GigaMUGA probe sequences against
  the mouse genome, I found the same number of non-unique markers
  (3,114), but they were not the same markers.

- Of the markers in common between the MegaMUGA and GigaMUGA arrays,
  the chromosome assignments have a lot of differences, even among
  those indicated to be unique.

In case that the probe sequences in the UNC annotation files might be
in error, I obtained the array probe sequences directly from GeneSeek.
These proved to be the same as those in the UNC files.

My goal here is to construct new annotation files for the
GigaMUGA and MegaMUGA arrays with
updated information about which markers map uniquely, which map to
multiple locations, and which do not map. I'll include genomic
coordinates only for the markers that map uniquely.

A summary of my findings:

- For markers mapping uniquely, the genomic coordinates in the UNC
  GigaMUGA annotation file were correct.

- The UNC MegaMUGA annotation file had coordinates in a previous mouse
  genome build (mm9). For markers mapping uniquely, there are a small
  number with differences in chromosome assignments; there are lots of
  differences in positions, with the new build (mm10).

- The "unique" column in the GigaMUGA annotation file is messed up.

- On both arrays, there are a number of pairs of markers with common probe
  sequences. Many of these are intended duplicates, but some are not
  and have inconsistent information

- Of the markers in common between the two arrays, some have different
  probe sequences on the two arrays, and there are cases where the
  sequence on one array maps uniquely but the other does not.

- I'm not sure what to do with the markers on the "P" chromosome,
  which I understand to be the pseudoautosomal region, as the markers
  are not mapping to the mm10 mouse genome build.



## Procedures

The key thing here was to run
[blastn](https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastSearch)
to map all array probe sequences to the mouse genome (build mm10).
There was some trickiness in trimming SNP alleles off some probes,
and in identifying the SNP position from the blast results.

[Separately, I studied the probe sequences in some
detail](https://kbroman.org/MUGAarrays/study_sequences).
Most probes do not include the SNP, which is the next base off the 3'
end. But for probes where `AlleleB_ProbeSeq` is provided (in addition
to `AlleleA_ProbeSeq`, which is always provided), the sequences
include the SNP allele as the last base. In these cases, I trimmed off
the SNP allele and used the 49-bp sequence.

I performed blast allowing no insertions/deletions. In cases where the
match didn't include the full probe sequence, I took all non-aligned
bases to be mismatches.

More precisely, I installed blastn on my linux laptop with

    sudo apt install ncbi-blast+"

I downloaded the mm10 mouse genome files from
<http://hgdownload-test.cse.ucsc.edu/goldenPath/mm10/bigZips/>

I created index files with

    makeblastdb -in MouseFasta/chr19.fa -dbtype nuc

I ran blastn with a command like this:

    blastn -db MouseFasta/chr19.fa -ungapped -out output_c19.txt -query gigamuga.fa -outfmt 6

For some reason, there were a small number of duplicate alignments
(same query, same chromosome, same start and end) in the results; I
omitted all but one of these duplicates.

I've also focused on full-length, perfect hits.


## Preliminaries

I'll first load some R packages.

```{r load_packages}
library(data.table)
library(broman)
library(devtools)
```

I'll now load various data sets: the GeneSeek and UNC annotation
files, and the blast results. For the blast results, I only keep the
full length, perfect hits.

```{r load_data}
# GeneSeek files
gm_gs <- data.table::fread("../GeneSeek/gigamuga_geneseek.csv",
                           skip=7, data.table=FALSE)
wh <- which(gm_gs[,1]=="[Controls]")
gm_gs <- gm_gs[1:(wh-1),]
rownames(gm_gs) <- gm_gs$Name

mm_gs <- data.table::fread("../GeneSeek/megamuga_geneseek.csv",
                           skip=7, data.table=FALSE)
wh <- which(mm_gs[,1]=="[Controls]")
mm_gs <- mm_gs[1:(wh-1),]
rownames(mm_gs) <- mm_gs$Name

# UNC files
load("../UNC/snps.gigamuga.Rdata")
gm_unc <- snps
# omit "chr" from the chromosome IDs
gm_unc$chr <- sub("chr", "", gm_unc$chr)

load("../UNC/snps.megamuga.Rdata")
mm_unc <- snps
# omit "chr" from the chromosome IDs
mm_unc$chr <- sub("chr", "", mm_unc$chr)

# Blast results; keep just the perfect matches(?)
mm_blast <- readRDS("../Blast/results_mm/mm_blastn_results.rds")
mm_blast <- mm_blast[mm_blast$tot_mismatch==0,]
gm_blast <- readRDS("../Blast/results_gm/gm_blastn_results.rds")
gm_blast <- gm_blast[gm_blast$tot_mismatch==0,]
```


## Summarize blast hits

I'll start with a bunch of code to summarize the blast hits. For each
SNP, I want to count the number of perfect blast hits and the number
of chromosomes hit. For markers with a unique hit, I'll record the
chromosome, position, and strand. For markers that hit multiple
locations but all on the same chromosome, I'll record the range of the
positions hit (in basepairs).

Here's the code for the MegaMUGA array:

```{r mm_summarize_blast}
# no. blast hits
mm_tab <- table(mm_blast$query)
mm_nchr <- mm_nhits <- setNames(rep(0, nrow(mm_gs)), rownames(mm_gs))
mm_nhits[names(mm_tab)] <- mm_tab

# no. chromosomes hit
mm_tab_chr <- table(mm_blast$query, mm_blast$chr)
mm_nchr[rownames(mm_tab_chr)] <- rowSums(mm_tab_chr > 0)

# chr,pos,strand for the unique ones
mm_blast_uniq <- mm_blast[mm_blast$query %in% names(mm_nhits)[mm_nhits==1],]
mm_blast_chr <- mm_blast_pos <- mm_blast_strand <- setNames(rep(NA, nrow(mm_gs)), rownames(mm_gs))
mm_blast_chr[mm_blast_uniq$query] <- mm_blast_uniq$chr
mm_blast_pos[mm_blast_uniq$query] <- mm_blast_uniq$snp_pos
mm_blast_strand[mm_blast_uniq$query] <- mm_blast_uniq$strand

# probe sequences
mm_blast_probe <- setNames(mm_gs$AlleleA_ProbeSeq, mm_gs$Name)
mm_trim <- (mm_gs$AlleleB_ProbeSeq != "")
mm_blast_probe[mm_trim] <- substr(mm_blast_probe[mm_trim], 1, nchar(mm_blast_probe[mm_trim])-1)

# SNP alleles
mm_blast_snp <- setNames(sub("]", "", mm_gs$SNP, fixed=TRUE), mm_gs$Name)
mm_blast_snp <- sub("[", "", mm_blast_snp, fixed=TRUE)
mm_blast_snp <- sub("/", "", mm_blast_snp, fixed=TRUE)

# put all of this stuff into a data frame
mm_uwisc <- data.frame(marker=names(mm_nhits),
                       n_blast_hits=mm_nhits,
                       n_blast_chr=mm_nchr,
                       unique=(mm_nhits==1),
                       multi=(mm_nhits>1),
                       no_pos=(mm_nhits==0),
                       chr=mm_blast_chr,
                       pos=mm_blast_pos,
                       strand=mm_blast_strand,
                       snp=mm_blast_snp[names(mm_nhits)],
                       probe=mm_blast_probe[names(mm_nhits)],
                       stringsAsFactors=FALSE)
rownames(mm_uwisc) <- mm_uwisc$marker

# probes that hit multiply but all to one chromosome
mm_mult_but1chr <- names(mm_nhits)[mm_nhits > 1 & mm_nchr==1]
mm_mult_but1chr_range <- sapply(mm_mult_but1chr, function(mar)
    diff(range(mm_blast$snp_pos[mm_blast$query==mar])))

# probes that hit uniquely on each of X and Y and to no other chromosome
mm_uniqXY <- names(mm_nhits)[rowSums(mm_tab_chr[,1:20])==0 & mm_tab_chr[,21]==1 & mm_tab_chr[,22]==1]
```


Now we repeat that for the GigaMUGA array:

```{r gm_summarize_blast}
# no. blast hits
gm_tab <- table(gm_blast$query)
gm_nchr <- gm_nhits <- setNames(rep(0, nrow(gm_gs)), rownames(gm_gs))
gm_nhits[names(gm_tab)] <- gm_tab

# no. chromosomes hit
gm_tab_chr <- table(gm_blast$query, gm_blast$chr)
gm_nchr[rownames(gm_tab_chr)] <- rowSums(gm_tab_chr > 0)

# chr,pos,strand for the unique ones
gm_blast_uniq <- gm_blast[gm_blast$query %in% names(gm_nhits)[gm_nhits==1],]
gm_blast_chr <- gm_blast_pos <- gm_blast_strand <- setNames(rep(NA, nrow(gm_gs)), rownames(gm_gs))
gm_blast_chr[gm_blast_uniq$query] <- gm_blast_uniq$chr
gm_blast_pos[gm_blast_uniq$query] <- gm_blast_uniq$snp_pos
gm_blast_strand[gm_blast_uniq$query] <- gm_blast_uniq$strand

# probe sequences
gm_blast_probe <- setNames(gm_gs$AlleleA_ProbeSeq, gm_gs$Name)
gm_trim <- (gm_gs$AlleleB_ProbeSeq != "")
gm_blast_probe[gm_trim] <- substr(gm_blast_probe[gm_trim], 1, nchar(gm_blast_probe[mm_trim])-1)

# SNP alleles
gm_blast_snp <- setNames(sub("]", "", gm_gs$SNP, fixed=TRUE), gm_gs$Name)
gm_blast_snp <- sub("[", "", gm_blast_snp, fixed=TRUE)
gm_blast_snp <- sub("/", "", gm_blast_snp, fixed=TRUE)

# put all of this stuff into a data frame
gm_uwisc <- data.frame(marker=names(gm_nhits),
                       n_blast_hits=gm_nhits,
                       n_blast_chr=gm_nchr,
                       unique=(gm_nhits==1),
                       multi=(gm_nhits>1),
                       no_pos=(gm_nhits==0),
                       chr=gm_blast_chr,
                       pos=gm_blast_pos,
                       strand=gm_blast_strand,
                       snp=gm_blast_snp[names(gm_nhits)],
                       probe=gm_blast_probe[names(gm_nhits)],
                       stringsAsFactors=FALSE)
rownames(gm_uwisc) <- gm_uwisc$marker

# probes that hit multiply but all to one chromosome
gm_mult_but1chr <- names(gm_nhits)[gm_nhits > 1 & gm_nchr==1]
gm_mult_but1chr_range <- sapply(gm_mult_but1chr, function(mar)
    diff(range(gm_blast$snp_pos[gm_blast$query==mar])))

# probes that hit uniquely on each of X and Y and to no other chromosome
gm_uniqXY <- names(gm_nhits)[rowSums(gm_tab_chr[,1:20])==0 & gm_tab_chr[,21]==1 & gm_tab_chr[,22]==1]
```




## MegaMUGA

### Basic blast results

I'll start by focusing on the MegaMUGA array, which contains
`r add_commas(nrow(mm_unc))`. In the UNC annotation file,
`r add_commas(sum(mm_unc$chr %in% (1:19)))` are on autosomes,
`r add_commas(sum(mm_unc$chr=="X"))` are on the X chromosome,
`r add_commas(sum(mm_unc$chr=="Y"))` are on the Y chromosome,
`r add_commas(sum(mm_unc$chr=="M"))` are in the mitochondrial genome,
and `r add_commas(sum(mm_unc$chr=="P"))` have chromosome "`P`"
(presumably pseudoautosomal).

In the blastn results, I find that `r add_commas(sum(mm_nhits==1))`
markers have a single, unique hit in the mouse genome (build mm10), while
`r add_commas(sum(mm_nhits>1))` hit to multiple locations
(median `r median(mm_nhits[mm_nhits>1])` locations and maximum
`r add_commas(max(mm_nhits))`), and `r add_commas(mm_nhits==0)`
have no perfect match in the mm10 build of the mouse genome.

### Unique markers

If we look at the inferred chromosome assignments of the markers with
a unique hit to the genome, all but
`r sum(mm_unc[mm_uwisc$marker, "chr"] != mm_uwisc$chr, na.rm=TRUE)`
match the chromosome in the UNC annotation file.

```{r mm_mismatch_chr}
# reorder mm_unc
mm_unc <- mm_unc[mm_uwisc$marker,]
wh <- mm_uwisc$unique & mm_uwisc$chr != mm_unc$chr
mm_diff_chr <- data.frame(marker=mm_uwisc$marker,
                          unc_chr=mm_unc$chr,
                          blast_chr=mm_uwisc$chr)[wh,]
mm_diff_chr <- mm_diff_chr[order(mm_diff_chr$marker),]
rownames(mm_diff_chr) <- 1:nrow(mm_diff_chr)
mm_diff_chr
```

Four of these were assigned to pseudoautosomal in the UNC annotations
but map to autosomes, one was mitochondrial (with a name to match),
but the sequence maps to chromosome1, and one (`MiniMuscle001`) was
assigned to chr 16 but maps to chr 11.

### Multiple hits to one chromosome

- markers that hit multiply but just to one chromosome


### Pseudoautosomal region


- chr P markers
  - no hits, unique, multiply
  - where do the unique ones hit?

  for MM, there are just 4 markers on chr P that are in the blast results
  ... and all 2 map uniquely to chr 13; the other two map uniquely to chr 5.
  NNT001, NNT002 but while these are chr P in MM, they're chr 13 in GM
  SAbGeoEUCOMM001 and SAbGeoEUCOMM002 are chr P in MM, map uniquely to
  chr 5, and former is on chr 5 in GM while the latter is not present on GM

- markers that hit X and Y uniquely, but no other chromosome
  - names
  - chr assignments in the UNC annotations


### What the heck is "`IlmnStrand`"?

- what's going on with the "strand" column in the GeneSeek annotation
  file. It has no apparent relationship to the strand of the hits.

- marker positions have no relationship to UNC file, but that's partly
  because the UNC file is in previous build

### Markers with the same probes

## GigaMUGA

I'll now turn to the GigaMUGA array, and do the same stuff plus a bit more.

the markers that are off by +/- 1 are exactly the ones with the AlleleB_ProbeSeq
(and I think I'm right)
```

```{r p_mar, eval=FALSE}
# markers on p chr in MM that are also in GM
pmar_mm <- mm_unc$marker[mm_unc$chr=="P" & mm_unc$marker %in% gm_unc$marker]

# markers on p chr(?) ...well actually chr is NA... in GM that are also in MM
pmar_gm <- gm_unc$marker[is.na(gm_unc$chr) & gm_unc$marker %in% mm_unc$marker]

pmar_gm %wnin% pmar_mm # UNC28724761 UNC28725371 UNC28726152
# latter are on 18 in MM
pmar_mm %wnin% pmar_gm # NNT001      NNT002      SAbGeoEUCOMM001
# latter are on 5,13 in GM
```

```{r markers_with_common_probes, eval=FALSE}
# MM probes with the same sequence
tab <- table(mm_unc$seq.A)
dup <- names(tab)[tab > 1] # 51 sets
z <- lapply(dup, function(d) { z <- mm_unc[mm_unc$seq.A == d,]
         if(length(unique(z$chr[!is.na(z$chr)])) > 1) return(rownames(z)) })
z <- z[sapply(z, length)>0]

# MM UNC88694 (chr 1), UNC2512583 (chr 2)
[only 2nd one in GM]
# MM UNC23833145 (chr 14), UNC4368900 (chr 2)
[in GM, both said to be unique, but same probe and on different chromosomes]
# MM UNC20251282 (chr 11), UNC5746160 (chr 3)
[in GM, both said to be unique, but same probe and on different chromosomes]
# MM UNC27690624 (chr 17), UNC30009421 (chr 19)
[only 2nd one in GM]
# MM UNC19598208 (chr 11)  UNC24941679 (chr 14)
[in GM, both said to be unique, but same probe and on different chromosomes]

# GM has *a lot* of markers with the same sequence
tab <- table(gm_unc$seq.A)
dup <- names(tab)[tab > 1] # 2672 sets
z <- lapply(dup, function(d) { z <- gm_unc[gm_unc$seq.A == d,]
         if(length(unique(z$chr[!is.na(z$chr)])) > 1) return(rownames(z)) })
z <- z[sapply(z, length)>0]
# only 10 sets that are on different chromosomes: 3 above, and these 7:
# UNCrs52300311 complement008 complement018 complement016 (repetitive sequence that doesn't map)
# Mit016 Mit030             # Mit016 placed on M
# sanger1454  sanger1454q   # both map multiply, but in unc annotations said unique
# UNC9964748  UNC10237349   # "
# sanger2519  sanger2519q
# sanger2310  sanger2310q
# JAX00187253q JAX00187253
```




## Session info

Here are the versions of R and R packages that I am using.

```{r session_info}
devtools::session_info()
```
