---
title:  "New MegaMUGA/GigaMUGA annotations"
author: "Karl Broman"
date:   "`r Sys.Date()`"
output:
    html_document:
        code_folding: hide
        toc: true
        toc_float: true
        toc_depth: 3
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.width=11, fig.height=6.5,
                      message=FALSE, warning=FALSE)
options(width=110)
```


## Introduction

While preparing a
[talk](https://kbroman.org/Talk_CTC2018/broman_ctc2018.pdf) and
[paper](https://www.biostat.wisc.edu/~kbroman/publications/mpp_diag.pdf)
on genotype diagnostics in Diversity Outbred mice,
I developed some concerns about the UNC annotation file for the
GigaMUGA array. In particular, I felt that some of the columns may
have been scrambled:

- In blasting the GigaMUGA probe sequences against
  the mouse genome, I found the same number of non-unique markers
  (3,114), but they were not the same markers.

- Of the markers in common between the MegaMUGA and GigaMUGA arrays,
  the chromosome assignments have a lot of differences, even among
  those indicated to be unique.

In case that the probe sequences in the UNC annotation files might be
in error, I obtained the array probe sequences directly from GeneSeek.
These proved to be the same as those in the UNC files.

My goal here is to construct new annotation files for the
GigaMUGA and MegaMUGA arrays with
updated information about which markers map uniquely, which map to
multiple locations, and which do not map. I'll include genomic
coordinates only for the markers that map uniquely.

A summary of my findings:

- For markers mapping uniquely, the genomic coordinates in the UNC
  GigaMUGA annotation file were correct.

- The UNC MegaMUGA annotation file had coordinates in a previous mouse
  genome build (mm9). For markers mapping uniquely, there are a small
  number with differences in chromosome assignments; there are lots of
  differences in positions, with the new build (mm10).

- The "unique" column in the GigaMUGA annotation file is messed up.

- On both arrays, there are a number of pairs of markers with common probe
  sequences. Many of these are intended duplicates, but some are not
  and have inconsistent information

- Of the markers in common between the two arrays, some have different
  probe sequences on the two arrays, and there are cases where the
  sequence on one array maps uniquely but the other does not.

- I'm not sure what to do with the markers on the "P" chromosome,
  which I understand to be the pseudoautosomal region, as the markers
  are not mapping to the mm10 mouse genome build.



## Procedures

The key thing here was to run
[blastn](https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastSearch)
to map all array probe sequences to the mouse genome (build mm10).
There was some trickiness in trimming SNP alleles off some probes,
and in identifying the SNP position from the blast results.

[Separately, I studied the probe sequences in some
detail](https://kbroman.org/MUGAarrays/study_sequences).
Most probes do not include the SNP, which is the next base off the 3'
end. But for probes where `AlleleB_ProbeSeq` is provided (in addition
to `AlleleA_ProbeSeq`, which is always provided), the sequences
include the SNP allele as the last base. In these cases, I trimmed off
the SNP allele and used the 49-bp sequence.

I performed blast allowing no insertions/deletions. In cases where the
match didn't include the full probe sequence, I took all non-aligned
bases to be mismatches.

More precisely, I installed blastn on my linux laptop with

    sudo apt install ncbi-blast+"

I downloaded the mm10 mouse genome files from
<http://hgdownload-test.cse.ucsc.edu/goldenPath/mm10/bigZips/>

I created index files with

    makeblastdb -in MouseFasta/chr19.fa -dbtype nuc

I ran blastn with a command like this:

    blastn -db MouseFasta/chr19.fa -ungapped -out output_c19.txt -query gigamuga.fa -outfmt 6

For some reason, there were a small number of duplicate alignments
(same query, same chromosome, same start and end) in the results; I
omitted all but one of these duplicates.

I've also focused on full-length, perfect hits.


## Preliminaries

I'll first load some R packages.

```{r load_packages}
library(data.table)
library(broman)
library(devtools)
```

I'll now load various data sets: the GeneSeek and UNC annotation
files, and the blast results. For the blast results, I only keep the
full length, perfect hits.

```{r load_data}
# GeneSeek files
gm_gs <- data.table::fread("../GeneSeek/gigamuga_geneseek.csv",
                           skip=7, data.table=FALSE)
wh <- which(gm_gs[,1]=="[Controls]")
gm_gs <- gm_gs[1:(wh-1),]
rownames(gm_gs) <- gm_gs$Name

mm_gs <- data.table::fread("../GeneSeek/megamuga_geneseek.csv",
                           skip=7, data.table=FALSE)
wh <- which(mm_gs[,1]=="[Controls]")
mm_gs <- mm_gs[1:(wh-1),]
rownames(mm_gs) <- mm_gs$Name

# UNC files
load("../UNC/snps.gigamuga.Rdata")
gm_unc <- snps
# omit "chr" from the chromosome IDs
gm_unc$chr <- sub("chr", "", gm_unc$chr)

load("../UNC/snps.megamuga.Rdata")
mm_unc <- snps
# omit "chr" from the chromosome IDs
mm_unc$chr <- sub("chr", "", mm_unc$chr)

# Blast results; keep just the perfect matches(?)
mm_blast <- readRDS("../Blast/results_mm/mm_blastn_results.rds")
mm_blast <- mm_blast[mm_blast$tot_mismatch==0,]
gm_blast <- readRDS("../Blast/results_gm/gm_blastn_results.rds")
gm_blast <- gm_blast[gm_blast$tot_mismatch==0,]
```


## Summarize blast hits

I'll start with a bunch of code to summarize the blast hits. For each
SNP, I want to count the number of perfect blast hits and the number
of chromosomes hit. For markers with a unique hit, I'll record the
chromosome, position, and strand. For markers that hit multiple
locations but all on the same chromosome, I'll record the range of the
positions hit (in basepairs).

Here's the code for the MegaMUGA array:

```{r mm_summarize_blast}
# no. blast hits
mm_tab <- table(mm_blast$query)
mm_nchr <- mm_nhits <- setNames(rep(0, nrow(mm_gs)), rownames(mm_gs))
mm_nhits[names(mm_tab)] <- mm_tab

# no. chromosomes hit
mm_tab_chr <- table(mm_blast$query, mm_blast$chr)
mm_nchr[rownames(mm_tab_chr)] <- rowSums(mm_tab_chr > 0)

# chr,pos,strand for the unique ones
mm_blast_uniq <- mm_blast[mm_blast$query %in% names(mm_nhits)[mm_nhits==1],]
mm_blast_chr <- mm_blast_pos <- mm_blast_strand <- setNames(rep(NA, nrow(mm_gs)), rownames(mm_gs))
mm_blast_chr[mm_blast_uniq$query] <- mm_blast_uniq$chr
mm_blast_pos[mm_blast_uniq$query] <- mm_blast_uniq$snp_pos
mm_blast_strand[mm_blast_uniq$query] <- mm_blast_uniq$strand

# probe sequences
mm_blast_probe <- setNames(mm_gs$AlleleA_ProbeSeq, mm_gs$Name)
mm_trim <- (mm_gs$AlleleB_ProbeSeq != "")
mm_blast_probe[mm_trim] <- substr(mm_blast_probe[mm_trim], 1, nchar(mm_blast_probe[mm_trim])-1)

# SNP alleles
mm_blast_snp <- setNames(sub("]", "", mm_gs$SNP, fixed=TRUE), mm_gs$Name)
mm_blast_snp <- sub("[", "", mm_blast_snp, fixed=TRUE)
mm_blast_snp <- sub("/", "", mm_blast_snp, fixed=TRUE)

# put all of this stuff into a data frame
mm_uwisc <- data.frame(marker=names(mm_nhits),
                       n_blast_hits=mm_nhits,
                       n_blast_chr=mm_nchr,
                       unique=(mm_nhits==1),
                       multi=(mm_nhits>1),
                       no_pos=(mm_nhits==0),
                       chr=mm_blast_chr,
                       pos=mm_blast_pos,
                       strand=mm_blast_strand,
                       snp=mm_blast_snp[names(mm_nhits)],
                       probe=mm_blast_probe[names(mm_nhits)],
                       stringsAsFactors=FALSE)
rownames(mm_uwisc) <- mm_uwisc$marker

# probes that hit multiply but all to one chromosome
mm_mult_but1chr <- names(mm_nhits)[mm_nhits > 1 & mm_nchr==1]
mm_mult_but1chr_range <- sapply(mm_mult_but1chr, function(mar)
    diff(range(mm_blast$snp_pos[mm_blast$query==mar])))

# probes that hit uniquely on each of X and Y and to no other chromosome
mm_hitXY <- rownames(mm_tab_chr)[rowSums(mm_tab_chr[,1:20])==0 & mm_tab_chr[,21]>0 & mm_tab_chr[,22]>0]
```


Now we repeat that for the GigaMUGA array:

```{r gm_summarize_blast}
# no. blast hits
gm_tab <- table(gm_blast$query)
gm_nchr <- gm_nhits <- setNames(rep(0, nrow(gm_gs)), rownames(gm_gs))
gm_nhits[names(gm_tab)] <- gm_tab

# no. chromosomes hit
gm_tab_chr <- table(gm_blast$query, gm_blast$chr)
gm_nchr[rownames(gm_tab_chr)] <- rowSums(gm_tab_chr > 0)

# chr,pos,strand for the unique ones
gm_blast_uniq <- gm_blast[gm_blast$query %in% names(gm_nhits)[gm_nhits==1],]
gm_blast_chr <- gm_blast_pos <- gm_blast_strand <- setNames(rep(NA, nrow(gm_gs)), rownames(gm_gs))
gm_blast_chr[gm_blast_uniq$query] <- gm_blast_uniq$chr
gm_blast_pos[gm_blast_uniq$query] <- gm_blast_uniq$snp_pos
gm_blast_strand[gm_blast_uniq$query] <- gm_blast_uniq$strand

# probe sequences
gm_blast_probe <- setNames(gm_gs$AlleleA_ProbeSeq, gm_gs$Name)
gm_trim <- (gm_gs$AlleleB_ProbeSeq != "")
gm_blast_probe[gm_trim] <- substr(gm_blast_probe[gm_trim], 1, nchar(gm_blast_probe[mm_trim])-1)

# SNP alleles
gm_blast_snp <- setNames(sub("]", "", gm_gs$SNP, fixed=TRUE), gm_gs$Name)
gm_blast_snp <- sub("[", "", gm_blast_snp, fixed=TRUE)
gm_blast_snp <- sub("/", "", gm_blast_snp, fixed=TRUE)

# put all of this stuff into a data frame
gm_uwisc <- data.frame(marker=names(gm_nhits),
                       n_blast_hits=gm_nhits,
                       n_blast_chr=gm_nchr,
                       unique=(gm_nhits==1),
                       multi=(gm_nhits>1),
                       no_pos=(gm_nhits==0),
                       chr=gm_blast_chr,
                       pos=gm_blast_pos,
                       strand=gm_blast_strand,
                       snp=gm_blast_snp[names(gm_nhits)],
                       probe=gm_blast_probe[names(gm_nhits)],
                       stringsAsFactors=FALSE)
rownames(gm_uwisc) <- gm_uwisc$marker

# probes that hit multiply but all to one chromosome
gm_mult_but1chr <- names(gm_nhits)[gm_nhits > 1 & gm_nchr==1]
gm_mult_but1chr_range <- sapply(gm_mult_but1chr, function(mar)
    diff(range(gm_blast$snp_pos[gm_blast$query==mar])))

# probes that hit uniquely on each of X and Y and to no other chromosome
gm_hitXY <- rownames(gm_tab_chr)[rowSums(gm_tab_chr[,1:20])==0 & gm_tab_chr[,21]>0 & gm_tab_chr[,22]>0]
```




## MegaMUGA

### Basic blast results

I'll start by focusing on the MegaMUGA array, which contains
`r add_commas(nrow(mm_unc))` markers. In the UNC annotation file,
`r add_commas(sum(mm_unc$chr %in% (1:19)))` are on autosomes,
`r add_commas(sum(mm_unc$chr=="X"))` are on the X chromosome,
`r add_commas(sum(mm_unc$chr=="Y"))` are on the Y chromosome,
`r add_commas(sum(mm_unc$chr=="M"))` are on the mitochondrial genome,
and `r add_commas(sum(mm_unc$chr=="P"))` have chromosome `P`, which
I take to mean they are pseudoautosomal.

In the blast results, I find that `r add_commas(sum(mm_nhits==1))`
markers have a single, unique hit in the mouse genome (build mm10), while
`r add_commas(sum(mm_nhits>1))` hit to multiple locations
(median `r median(mm_nhits[mm_nhits>1])` locations and maximum
`r add_commas(max(mm_nhits))`), and `r add_commas(sum(mm_nhits==0))`
have no perfect match in the mm10 build of the mouse genome.

### Unique markers

If we look at the inferred chromosome assignments of the markers with
a unique hit to the genome, all but
`r sum(mm_unc[mm_uwisc$marker, "chr"] != mm_uwisc$chr, na.rm=TRUE)`
match the chromosome in the UNC annotation file. Here are the markers
with discrepancies in chromosome assignment:

```{r mm_mismatch_chr}
# reorder mm_unc
mm_unc <- mm_unc[mm_uwisc$marker,]
wh <- mm_uwisc$unique & mm_uwisc$chr != mm_unc$chr
mm_diff_chr <- data.frame(marker=mm_uwisc$marker,
                          unc_chr=mm_unc$chr,
                          blast_chr=mm_uwisc$chr,
                          stringsAsFactors=FALSE)[wh,]
mm_diff_chr <- mm_diff_chr[order(mm_diff_chr$marker),]
rownames(mm_diff_chr) <- 1:nrow(mm_diff_chr)
mm_diff_chr
```

Four of these were pseudoautosomal in the UNC annotations
but map to autosomes, one was mitochondrial (with a name to match)
but the sequence maps to chr 1, and one (`MiniMuscle001`) was
assigned to chr 16 but maps to chr 11.

`r Numbers[sum(mm_diff_chr$marker %in% gm_unc$marker)]`
of these `r numbers[nrow(mm_diff_chr)]`
markers are found on the GigaMUGA array (all but
`r mm_diff_chr$marker %wnin% gm_unc$marker`), and
the chromosome assignments in the UNC annotations of the GigaMUGA
array all match what we see in these blast results, rather than the
chromosome in the UNC annotations of the MegaMUGA array.

I won't make any comparisons of basepair positions, as the UNC
annotation file for the MegaMUGA array uses coordinates from build mm9
of the mouse genome, rather than build mm10 that I'm using.

### Multiple hits to one chromosome

There are `r length(mm_mult_but1chr)` markers on the MegaMUGA array
that have multiple perfect hits in the mouse genome, but all on the
same chromosome. I looked at the range in SNP positions, for these
multiple hits, which varied from
`r min(mm_mult_but1chr_range)` bp to
`r round(max(mm_mult_but1chr_range)/1e6)` Mbp, with a median of
`r round(median(mm_mult_but1chr_range)/1000)` kbp.

There are `r sum(mm_mult_but1chr_range < 1000)` markers with multiple
hits but all on a single chromosome and all within a 1000 bp interval.
We could maybe keep these, and assign each to the median of the observed
positions.  But for now we'll leave them with missing genomic
coordinates.


### Pseudoautosomal region

I'm not sure what to do about the pseudoautosomal markers.
The UNC annotations for the MegaMUGA array has
`r sum(mm_unc$chr=="P")` markers with chromosome `P`. But only
`r numbers[sum(mm_uwisc$unique[mm_unc$chr=="P"])]`
have hits to the mm10 mouse genome build, and these are the
four mentioned above, that actually map to autosomes. The other
`r sum(mm_uwisc$no_pos[mm_unc$chr=="P"])` have no perfect hits in the
blast results, which leads me to think that the pseudoautosomal region
is perhaps not part of the mm10 genome assembly.

There are another `r sum(grepl("^PAR[0-9]+$", mm_unc$marker))` markers
with names like `PAR10`. All of these were placed on chr X in the UNC
annotation file. In the blast results,
`r numbers[sum(mm_uwisc$unique[grepl("^PAR[0-9]+$", mm_uwisc$marker)])]`
of these have unique hits in the mouse genome, all to chromosome X.
The remainder have multiple hits.
PAR4 has a single hit on each of the X and Y chromosomes.
PAR6 has a single hit on the X chromosome and two hits on the Y
chromosome.
The other `r sum(grep("^PAR[0-9]+$", mm_unc$marker, value=TRUE) %in% mm_mult_but1chr)`
PAR markers have multiple hits but just to the X chromosome.

There are two additional markers with hits to both the X and the Y
chromosome, but with no hits to autosomes:
`r vec2string(mm_hitXY[mm_hitXY %nin% c("PAR4", "PAR6")])`.
UNC31596388 has a single hit on each of the X and Y chromosomes.
UNC31596457 has a single hit on the X and two hits on the Y.
So maybe these are also pseudoautosomal markers.


### What the heck is `IlmnStrand`?

The annotation file I received from GeneSeek has a column `IlmnStrand`
with values `TOP` and `BOT`, which I took to indicate the strand
origin of the SNP probe sequences (which I thought were generally
called "plus" and "minus").

But while this `IlmnStrand` column is consistent with the
`SourceStrand` column in the UNC annotation file, it doesn't have any
relation to the strands that I'm seeing in the blast results.

About half of the `IlmnStrand` values are `TOP` and half are `BOT`,
and in each of these groups, markers with a unique hit to the mouse
genome are about half hitting the plus strand and half hitting the
minus strand, so I don't know what to make of this.



### Markers with the same probes

One last thing on the MegaMUGA array: there are
`r sum((mm_probe_tab <- table(mm_uwisc$probe))==2)` pairs of markers
with the same probe sequence, and one set of four markers that all of
the same sequence.

For most of these replicate problems, it seems like they were
intended, as they have names like JAX00333860 and repJAX00333860,
or CEAJAX00159275 and CEBJAX00159275.

But the group of four probes with matching sequences are
`r vec2string(sort(mm_uwisc$marker[mm_uwisc$probe==names(mm_probe_tab)[mm_probe_tab==4]]))`,
all on the Y chromosome. So two pairs of intended replicates, but it's
not obvious that JAX00725045 and JAX00725054 should be the same.

Of the other `r sum(mm_probe_tab==2)` pairs, 35 are obvious from their
names. Here are the others:

```{r mm_study_dup_pairs}
dup <- names(mm_probe_tab)[mm_probe_tab==2]
dup_pairs <- t(sapply(dup, function(a) mm_uwisc$marker[mm_uwisc$probe==a]))
dup_pairs <- dup_pairs[dup_pairs[,2] != paste0("rep", dup_pairs[,1]) &
                       dup_pairs[,1] != paste0("rep", dup_pairs[,2]) &
                       dup_pairs[,2] != paste0("CEA", dup_pairs[,2]) &
                       dup_pairs[,1] != paste0("CEA", dup_pairs[,2]) &
                       sub("^CEA", "CEB", dup_pairs[,2]) != dup_pairs[,1] &
                       sub("^CEA", "CEB", dup_pairs[,1]) != dup_pairs[,2],]
dup_pairs <- t(apply(dup_pairs, 1, sort))
rownames(dup_pairs) <- 1:nrow(dup_pairs)
colnames(dup_pairs) <- c("snp1", "snp2")
dup_pairs
```

This begs the question: are there other marker pairs whose names
indicate that they should be duplicates, but where they're not really
duplicates?

```{r mm_look_for_more_dups}
mn <- mm_uwisc$marker
mnrep <- paste0("rep", mn)
mncea <- paste0("CEA", mn)
mnceb <- paste0("CEB", substr(mn, 4, nchar(mn)))
mnceb[grepl("^CEB", mn)] <- "" # drop those that already start with CEB

more_dup <- rbind(cbind(mn, mnrep)[mnrep %in% mn,],
                  cbind(mn, mncea)[mncea %in% mn,],
                  cbind(mn, mnceb)[mnceb %in% mn,])

more_dup_probes <- cbind(mm_uwisc[more_dup[,1],"probe"],
                         mm_uwisc[more_dup[,2],"probe"])

stopifnot(more_dup_probes[,1] == more_dup_probes[,2])
```

If I look for the patterns that I saw in the marker pairs with common
sequences, I see a total of `r nrow(more_dup)` pairs, but these are
exactly the ones found based on their sequences matching (including
the two pairs that make up the set of four markers all with the same
sequence).




## GigaMUGA

I'll now turn to the GigaMUGA array, and do the same stuff plus a bit more.

the markers that are off by +/- 1 are exactly the ones with the AlleleB_ProbeSeq
(and I think I'm right)
```

```{r p_mar, eval=FALSE}
# markers on p chr in MM that are also in GM
pmar_mm <- mm_unc$marker[mm_unc$chr=="P" & mm_unc$marker %in% gm_unc$marker]

# markers on p chr(?) ...well actually chr is NA... in GM that are also in MM
pmar_gm <- gm_unc$marker[is.na(gm_unc$chr) & gm_unc$marker %in% mm_unc$marker]

pmar_gm %wnin% pmar_mm # UNC28724761 UNC28725371 UNC28726152
# latter are on 18 in MM
pmar_mm %wnin% pmar_gm # NNT001      NNT002      SAbGeoEUCOMM001
# latter are on 5,13 in GM
```

```{r markers_with_common_probes, eval=FALSE}
# MM probes with the same sequence
tab <- table(mm_unc$seq.A)
dup <- names(tab)[tab > 1] # 51 sets
z <- lapply(dup, function(d) { z <- mm_unc[mm_unc$seq.A == d,]
         if(length(unique(z$chr[!is.na(z$chr)])) > 1) return(rownames(z)) })
z <- z[sapply(z, length)>0]

# MM UNC88694 (chr 1), UNC2512583 (chr 2)
[only 2nd one in GM]
# MM UNC23833145 (chr 14), UNC4368900 (chr 2)
[in GM, both said to be unique, but same probe and on different chromosomes]
# MM UNC20251282 (chr 11), UNC5746160 (chr 3)
[in GM, both said to be unique, but same probe and on different chromosomes]
# MM UNC27690624 (chr 17), UNC30009421 (chr 19)
[only 2nd one in GM]
# MM UNC19598208 (chr 11)  UNC24941679 (chr 14)
[in GM, both said to be unique, but same probe and on different chromosomes]

# GM has *a lot* of markers with the same sequence
tab <- table(gm_unc$seq.A)
dup <- names(tab)[tab > 1] # 2672 sets
z <- lapply(dup, function(d) { z <- gm_unc[gm_unc$seq.A == d,]
         if(length(unique(z$chr[!is.na(z$chr)])) > 1) return(rownames(z)) })
z <- z[sapply(z, length)>0]
# only 10 sets that are on different chromosomes: 3 above, and these 7:
# UNCrs52300311 complement008 complement018 complement016 (repetitive sequence that doesn't map)
# Mit016 Mit030             # Mit016 placed on M
# sanger1454  sanger1454q   # both map multiply, but in unc annotations said unique
# UNC9964748  UNC10237349   # "
# sanger2519  sanger2519q
# sanger2310  sanger2310q
# JAX00187253q JAX00187253
```




## Session info

Here are the versions of R and R packages that I am using.

```{r session_info}
devtools::session_info()
```
