---
title:  "New MegaMUGA/GigaMUGA annotations"
author: "Karl Broman"
date:   "`r Sys.Date()`"
output:
    html_document:
        code_folding: hide
        toc: true
        toc_float: true
        toc_depth: 3
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.width=11, fig.height=6.5,
                      message=FALSE, warning=FALSE)
options(width=110)
```


## Introduction

While preparing a
[talk](https://kbroman.org/Talk_CTC2018/broman_ctc2018.pdf) and
[paper](https://www.biostat.wisc.edu/~kbroman/publications/mpp_diag.pdf),
I developed some concerns about the UNC annotation file for the
GigaMUGA array. In particular, I felt that some of the columns may
have been scrambled:

- In blasting the GigaMUGA probe sequences against
  the mouse genome, I found the same number of non-unique markers
  (3,114), but they were not the same markers. In comparing

- Of the markers in common between the MegaMUGA and GigaMUGA arrays,
  the chromosome assignments have a lot of differences.

In case that the probe sequences in the UNC annotation files might be
in error, I obtained the array probe sequences directly from GeneSeek.
These proved to be the same as those in the UNC files.

My main goal has been to construct new annotation files for the
GigaMUGA and MegaMUGA arrays, with consistency across the two, with
updated information about which markers map uniquely, which map to
multiple locations, and which do not map, and to include genomic
coordinates only for the markers that map uniquely.

A summary of my findings:

- For markers mapping uniquely, the genomic coordinates in the UNC
  GigaMUGA annotation file were correct.

- The UNC MegaMUGA annotation file had coordinates in a previous mouse
  genome build (mm9). For markers mapping uniquely, there are a small
  number with differences in chromosome assignments; there are lots of
  differences in positions, with the new build (mm10).

- The "unique" column in the GigaMUGA annotation file is messed up.


## Procedures

The key thing here was to run blastn to map all array probe sequences
to the mouse genome (build mm10). There is some trickiness in
trimming SNP alleles off some problems, and in identifying the SNP
position from the blast results.


## Preliminaries

I'll first load some R packages.

```{r load_packages}
library(data.table)
library(broman)
library(devtools)
```

I'll now load various data sets: the GeneSeek and UNC annotation
files, and the blast results.

```{r load_data}
# GeneSeek files
gm_gs <- data.table::fread("../GeneSeek/gigamuga_geneseek.csv",
                           skip=7, data.table=FALSE)
wh <- which(gm_gs[,1]=="[Controls]")
gm_gs <- gm_gs[1:(wh-1),]
rownames(gm_gs) <- gm_gs$Name

mm_gs <- data.table::fread("../GeneSeek/megamuga_geneseek.csv",
                           skip=7, data.table=FALSE)
wh <- which(mm_gs[,1]=="[Controls]")
mm_gs <- mm_gs[1:(wh-1),]
rownames(mm_gs) <- mm_gs$Name

# UNC files
load("../UNC/snps.gigamuga.Rdata")
gm_unc <- snps

load("../UNC/snps.megamuga.Rdata")
mm_unc <- snps

# Blast results; keep just the perfect matches(?)
mm_blast <- readRDS("../Blast/results_mm/mm_blastn_results.rds")
mm_blast <- mm_blast[mm_blast$n_mismatch==0,]
gm_blast <- readRDS("../Blast/results_gm/gm_blastn_results.rds")
gm_blast <- gm_blast[gm_blast$n_mismatch==0,]
```


## Study blast results

```{r summarize_blast_results}
# no. blast hits
gm_tab <- table(gm_blast$query)
gm_nchr <- gm_nhits <- setNames(rep(0, nrow(gm_gs)), rownames(gm_gs))
gm_nhits[names(gm_tab)] <- gm_tab

# no. chromosomes hit
gm_tab_chr <- table(gm_blast$query, gm_blast$chr)
gm_nchr[rownames(gm_tab_chr)] <- rowSums(gm_tab_chr > 0)

## Same for MegaMUGA
# no. blast hits
mm_tab <- table(mm_blast$query)
mm_nchr <- mm_nhits <- setNames(rep(0, nrow(mm_gs)), rownames(mm_gs))
mm_nhits[names(mm_tab)] <- mm_tab

# no. chromosomes hit
mm_tab_chr <- table(mm_blast$query, mm_blast$chr)
mm_nchr[rownames(mm_tab_chr)] <- rowSums(mm_tab_chr > 0)
```


## Session info

Here are the versions of R and R packages that I am using.

```{r session_info}
devtools::session_info()
```
