---
title:  "Study SNP sequences on MegaMUGA and GigaMUGA arrays"
author: "Karl Broman"
date:   "`r Sys.Date()`"
output:
    html_document:
        code_folding: hide
        toc: true
        toc_float: true
        toc_depth: 3
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.width=11, fig.height=6.5,
                      message=FALSE, warning=FALSE)
options(width=110)
```


## Preliminaries

I'll first load some R packages.

```{r load_packages}
library(data.table)
library(broman)
```

I'll now load the GeneSeek annotation files for the MegaMUGA and
GigaMUGA arrays. There are a few rows above the header, and then some control
rows at the bottom, that need to be clipped out.

```{r load_geneseek_files}
gm <- data.table::fread("../GeneSeek/gigamuga_geneseek.csv",
                        skip=7, data.table=FALSE)
wh <- which(gm[,1]=="[Controls]")
gm <- gm[1:(wh-1),]

mm <- data.table::fread("../GeneSeek/megamuga_geneseek.csv",
                        skip=7, data.table=FALSE)
wh <- which(mm[,1]=="[Controls]")
mm <- mm[1:(wh-1),]
```

I'll also load the file with the names that are in common between the
two arrays.

```{r load_common_markers}
common <- data.table::fread("../GeneSeek/common_markers.csv", data.table=FALSE)
common_markers <- common[-1,1]
```

## Numbers of markers

The GigaMUGA array contains `r add_commas(nrow(gm))` markers, which
the MegaMUGA has `r add_commas(nrow(mm))` markers.
The marker names in each file
`r ifelse(length(unique(mm$Name)) == nrow(mm) && length(unique(gm$Name)) == nrow(gm), "are", "are not")`
all distinct.
There are `r add_commas(sum(mm$Name %in% gm$Name))` markers in common between
the two arrays.

The Excel file that I received from GeneSeek has three worksheets;
the second and third worksheets contain annotation information, including probe
sequences, for the GigaMUGA and MegaMUGA files, respectively. The first worksheet
indicates which markers are in common between the GigaMUGA and MegaMUGA arrays,
but it's a bit odd. It has counts of markers on each array, which
`r ifelse(common[1,4] == nrow(gm) && common[2,4] == nrow(mm), "are", "are not")`
correct, but there's a count that says "`On Both`",
but it's `r add_commas(common[3,4])`, which is different that what you get in a
direct check of the marker names. Also it has a list of names,
presumably the ones that are in common, and it does have
`r add_commas(nrow(common)-1)` markers, but only there are only
`r add_commas(length(unique(common_markers)))` distinct names, and only
`r add_commas(sum(common_markers %in% gm$Name))` are present in the GigaMUGA
array (`r ifelse(sum(common_markers %in% gm$Name) == sum(common_markers %in% mm$Name), "The same", "A different")`,
number of markers are in the MegaMUGA array.)
The other `r sum(common_markers %nin% gm$Name)` markers contain all
the duplicates and look like control stuff and not real marker names:
`r vec2string(unique(common_markers %wnin% gm$Name))`.
It seems like those came from the `[Control]` section of the other two
worksheets.

The key summaries here:

- GigaMUGA has `r add_commas(nrow(gm))` SNPs
- MegaMUGA has `r add_commas(nrow(mm))` SNPs
- There are `r add_commas(sum(mm$Name %in% gm$Name))` in common


## Basic sequence info

The GeneSeek annotation files contain four sets of sequences.
`AlleleA_ProbeSeq`, `AlleleB_ProbeSeq`, `SourceSeq`, and `TopGenomicSeq`.

The `AlleleA_ProbeSeq` sequences are
`r ifelse(all(nchar(c(gm$AlleleA_ProbeSeq, mm$AlleleA_ProbeSeq))==50), "all", "*not* all")`
length 50.

The `AlleleB_ProbeSeq` sequences are mostly missing. In the GigaMUGA file there
are `r sum(gm$AlleleB_ProbeSeq != "")` that are not missing; in the MegaMUGA file there
are `r sum(mm$AlleleB_ProbeSeq != "")` that are not missing.

The `SourceSeq` and `TopGenomicSeq` sequences are the same
information. They're identical when the `IlmnStrand` column is `TOP`.
The latter is the reverse complement of the former when the
`IlmnStrand` column is `BOT`.

The `SourceSeq` all seem to contain text like `[A/C]` that indicates
the SNP. This is absent from the `AlleleA_ProbeSeq` which are all
strictly ACGT characters.

## extra stuff

- lengths of the sequences
- is it obvious which base is the SNP?
- markers with identical sequences
- markers in common between the two arrays ... are the sequences the same?
- relationship between probe sequence and source sequence
- what to do with the "IlmnStrand" and the SNP allele?
- seems like snp is just *after* the probe sequence
  sometimes probe B has sequence; seems like then the probe sequence includes the snp
  probe sequences are always 50 bp

I guess check:
- always SourceSeq == TopGenomicSeq?
- when allele B probe sequence present, the sequences are identical except for last base,
  (and that last base is the SNP allele)
