---
title:  "Study SNP sequences on MegaMUGA and GigaMUGA arrays"
author: "Karl Broman"
date:   "`r Sys.Date()`"
output:
    html_document:
        code_folding: hide
        toc: true
        toc_float: true
        toc_depth: 3
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.width=11, fig.height=6.5,
                      message=FALSE, warning=FALSE)
options(width=110)
```


## Preliminaries

I'll first load some R packages.

```{r load_packages}
library(data.table)
library(broman)
library(devtools)
```

I'll now load the GeneSeek annotation files for the MegaMUGA and
GigaMUGA arrays. There are a few rows above the header, and then some control
rows at the bottom, that need to be clipped out.

```{r load_geneseek_files}
gm <- data.table::fread("../GeneSeek/gigamuga_geneseek.csv",
                        skip=7, data.table=FALSE)
wh <- which(gm[,1]=="[Controls]")
gm <- gm[1:(wh-1),]
rownames(gm) <- gm$Name

mm <- data.table::fread("../GeneSeek/megamuga_geneseek.csv",
                        skip=7, data.table=FALSE)
wh <- which(mm[,1]=="[Controls]")
mm <- mm[1:(wh-1),]
rownames(mm) <- mm$Name
```

I'll also load the file with the names that are in common between the
two arrays.

```{r load_common_markers}
common <- data.table::fread("../GeneSeek/common_markers.csv", data.table=FALSE)
common_markers <- common[-1,1]
```

## Numbers of markers

The GigaMUGA array contains `r add_commas(nrow(gm))` markers, which
the MegaMUGA has `r add_commas(nrow(mm))` markers.
The marker names in each file
`r ifelse(length(unique(mm$Name)) == nrow(mm) && length(unique(gm$Name)) == nrow(gm), "are", "are not")`
all distinct.
There are `r add_commas(sum(mm$Name %in% gm$Name))` markers in common between
the two arrays.

The Excel file that I received from GeneSeek has three worksheets;
the second and third worksheets contain annotation information, including probe
sequences, for the GigaMUGA and MegaMUGA files, respectively. The first worksheet
indicates which markers are in common between the GigaMUGA and MegaMUGA arrays,
but it's a bit odd. It has counts of markers on each array, which
`r ifelse(common[1,4] == nrow(gm) && common[2,4] == nrow(mm), "are", "are not")`
correct, but there's a count that says "`On Both`",
but it's `r add_commas(common[3,4])`, which is different that what you get in a
direct check of the marker names. Also it has a list of names,
presumably the ones that are in common, and it does have
`r add_commas(nrow(common)-1)` markers, but only there are only
`r add_commas(length(unique(common_markers)))` distinct names, and only
`r add_commas(sum(common_markers %in% gm$Name))` are present in the GigaMUGA
array (`r ifelse(sum(common_markers %in% gm$Name) == sum(common_markers %in% mm$Name), "The same", "A different")`,
number of markers are in the MegaMUGA array.)
The other `r sum(common_markers %nin% gm$Name)` markers contain all
the duplicates and look like control stuff and not real marker names:
`r vec2string(unique(common_markers %wnin% gm$Name))`.
It seems like those came from the `[Control]` section of the other two
worksheets.

The key summaries here:

- GigaMUGA has `r add_commas(nrow(gm))` SNPs
- MegaMUGA has `r add_commas(nrow(mm))` SNPs
- There are `r add_commas(sum(mm$Name %in% gm$Name))` in common


## Available sequences

The GeneSeek annotation files contain four sets of sequences.
`AlleleA_ProbeSeq`, `AlleleB_ProbeSeq`, `SourceSeq`, and `TopGenomicSeq`.

The `AlleleA_ProbeSeq` sequences are
`r ifelse(all(nchar(c(gm$AlleleA_ProbeSeq, mm$AlleleA_ProbeSeq))==50), "all", "*not* all")`
length 50.

The `AlleleB_ProbeSeq` sequences are mostly missing. In the GigaMUGA file there
are `r sum(gm$AlleleB_ProbeSeq != "")` that are not missing; in the MegaMUGA file there
are `r sum(mm$AlleleB_ProbeSeq != "")` that are not missing.

The `SourceSeq` and `TopGenomicSeq` sequences are the same
information. They're identical when the `IlmnStrand` column is `TOP`.
The latter is the reverse complement of the former when the
`IlmnStrand` column is `BOT`.

The `SourceSeq` all seem to contain text like `[A/C]` that indicates
the SNP. This is absent from the `AlleleA_ProbeSeq` which are all
strictly ACGT characters.

## Duplicate probe sequences

There probe sequences are not all distinct. On the GigaMUGA array,
there are `r add_commas(sum(table(gm$AlleleA_ProbeSeq)==2))` that
appear twice, `r sum(table(gm$AlleleA_ProbeSeq)==3)` that appear three
times, `r sum(table(gm$AlleleA_ProbeSeq)==4)` that appear four times.

On the MegaMUGA array,
there are `r add_commas(sum(table(mm$AlleleA_ProbeSeq)==2))` that
appear twice, `r sum(table(mm$AlleleA_ProbeSeq)==3)` that appear three
times, `r sum(table(mm$AlleleA_ProbeSeq)==4)` that appears four times.

## Probe sequences for common markers

Of the `r add_commas(length(common <- mm$Name %win% gm$Name))` markers
that are in common between the two arrays, there are
`r sum(mm[common,"AlleleA_ProbeSeq"] != gm[common, "AlleleA_ProbeSeq"])`
markers that have different probe sequences.

- All cases where one is top strand and other is bottom strand?
- And so one takes sourceseq to left of snp and other to right of snp?
  (because in all cases the snp allele is at end or just after the
  probe sequence?)

## AlleleB probe sequences

For the small number of cases where there's an `AlleleB_ProbeSeq`, in
addition to `AlleleA_ProbeSeq`, I think the last base on the two is
the SNP. They should match at the other 49 bases. Here's a quick check
of that:

```{r check_alleleb_probes}
# cases with AlleleB probe sequence
wh_gm <- which(gm$AlleleB_ProbeSeq != "")
# grab the two sequences
probeA <- gm$AlleleA_ProbeSeq[wh_gm]
probeB <- gm$AlleleB_ProbeSeq[wh_gm]
# first 49 characters are the same
stopifnot( all(substr(probeA, 1, 49) == substr(probeB, 1, 49)) )
# grab the last base from each, which should be the alleles
alleleA <- substr(probeA, 50, 50)
alleleB <- substr(probeB, 50, 50)
# the alleles are different
stopifnot( all( alleleA != alleleB ) )
# the alleles match the SNP column
stopifnot( all( gm$SNP[wh_gm] == paste0("[", alleleA, "/", alleleB, "]") ) )

# now all that for the MegaMUGA array...
# cases with AlleleB probe sequence
wh_mm <- which(mm$AlleleB_ProbeSeq != "")
# grab the two sequences
probeA <- mm$AlleleA_ProbeSeq[wh_mm]
probeB <- mm$AlleleB_ProbeSeq[wh_mm]
# first 49 characters are the same
stopifnot( all(substr(probeA, 1, 49) == substr(probeB, 1, 49)) )
# grab the last base from each, which should be the alleles
alleleA <- substr(probeA, 50, 50)
alleleB <- substr(probeB, 50, 50)
# the alleles are different
stopifnot( all( alleleA != alleleB ) )
# the alleles match the SNP column
stopifnot( all( mm$SNP[wh_mm] == paste0("[", alleleA, "/", alleleB, "]") ) )
```

The code aboved checked that for the cases where there's an AlleleB
probe sequence, the first 49 bases are the same, between the two
alleles, while the 50th base is the SNP

For the remaining probes, I believe that the SNPs is just after the
probe sequence. We'll check that by comparing it to the source
sequences, next.

## Compare probes to source sequences

I'm presuming that the probe sequences match the part of the source
sequence from just before the `[A/T]` SNP part, but I want to check.

I'll first check that the `SourceSeq` column matches the
`TopGenomicSeq` column, using the reverse complement when
`SourceStrand` is `"BOT"`.

```{r compare_sourceseq_topgenomicseq}
source("revcomp.R") # reverse complement

# reverse complement of source sequence
#    ...a few cases of lower case letters in the latter
#    ...and it's tricky handling the SNP alleles, b/c they shouldn't be swapped
gm_source_rev <- revcomp(gm$SourceSeq)

# when SourceStrand=="TOP", the SourceSeq and TopGenomicSeq match
# when SourceStrand=="BOT", the TopGenomicSeq is the reverse complement
stopifnot( all((gm$SourceSeq == gm$TopGenomicSeq & gm$SourceStrand=="TOP") |
               (gm_source_rev == gm$TopGenomicSeq & gm$SourceStrand=="BOT")) )

# now same thing with MegaMUGA
# reverse complement of source sequence
#    ...a few cases of lower case letters in the latter
#    ...and it's tricky handling the SNP alleles, b/c they shouldn't be swapped
mm_source_rev <- revcomp(mm$SourceSeq)

# when SourceStrand=="TOP", the SourceSeq and TopGenomicSeq match
# when SourceStrand=="BOT", the TopGenomicSeq is the reverse complement
stopifnot( all((mm$SourceSeq == mm$TopGenomicSeq & mm$SourceStrand=="TOP") |
               (mm_source_rev == mm$TopGenomicSeq & mm$SourceStrand=="BOT")) )
```

If we made it here, than that was true and the `SourceSeq` and `TopGenomicSeq`
columns have the same information, so we can just focus on the `SourceSeq` column.

I now want to compare the SNP information in the source sequence to
the SNP column.


```{r compare_SourceSeq_to_SNP}
# grab SNP info from SourceSeq, with alleles in both orders
gm_source_snp <- sub("^.*\\[", "[", sub("\\].*$", "]", gm$SourceSeq))
gm_source_snp_comp <- justcomp(gm_source_snp)

# check that SNP column is one or the other
stopifnot( all(gm_source_snp == gm$SNP | gm_source_snp_comp == gm$SNP) )

# BUT NOTE: I thought whether it was one or the other depended on whether
#           IlmnStrand == SourceStrand or not, but there are some differences.
gm_snp_strand_problem <- ((gm_source_snp == gm$SNP & gm$IlmnStrand != gm$SourceStrand) |
                          (gm_source_snp_comp == gm$SNP & gm$IlmnStrand == gm$SourceStrand))

# but these are all A/T or C/G cases, so just about defining which allele is A vs B
stopifnot( all( gm_source_snp[gm_snp_strand_problem] %in% c("[A/T]", "[C/G]") ) )

mm_source_snp <- sub("^.*\\[", "[", sub("\\].*$", "]", mm$SourceSeq))
mm_source_snp_comp <- justcomp(mm_source_snp)

# check that SNP column is one or the other
stopifnot( all(mm_source_snp == mm$SNP | mm_source_snp_comp == mm$SNP) )

# mismatches between direct/complement match and whether strands are the same
mm_snp_strand_problem <- ((mm_source_snp == mm$SNP & mm$IlmnStrand != mm$SourceStrand) |
                          (mm_source_snp_comp == mm$SNP & mm$IlmnStrand == mm$SourceStrand))

# but these are all A/T or C/G cases, so just about defining which allele is A vs B
stopifnot( all( mm_source_snp[mm_snp_strand_problem] %in% c("[A/T]", "[C/G]") ) )
```

From this, we find that the `SNP` column is the same as the SNP in the
`SourceSeq`, or it is the complement of that. I thought whether it was
a direct match or a match to the complement would be according to
whether `IlmnStrand == SourceStrand` or not, but that there are `r
sum(gm_snp_strand_problem)` mismatches on the GigaMUGA and `r
sum(mm_snp_strand_problem)` mismatches on the MegaMUGA. But these are
all cases of A/T or C/G SNPs, so it's just a matter of the definition
of the A vs B alleles.

Finally (for this section), I'll now turn to whether the SNP sequences match the source sequences.

I'll first split the source sequences into the bits before and after
the snp. I'll also compare the SNP part of the source sequence to the
SNP column. Then I'll compare the parts of the source sequence to
AlleleA probes.

```{r probe_seq_vs_source_seq}
# grab the sequence before and after the SNP
gm_source_spl <- strsplit(gm$SourceSeq, "\\[[ACGT]/[ACGT]\\]")
gm_before_snp <- sapply(gm_source_spl, "[", 1) # a bunch of these have length 0
gm_after_snp <- sapply(gm_source_spl, "[", 2)
gm_after_snp[is.na(gm_after_snp)] <- "" # turn NAs into emptys

# for the SNPs that have both A and B probe sequences, drop the last character from the A sequence
gm_probe <- gm$AlleleA_ProbeSeq
gm_probe[gm$AlleleB_ProbeSeq != ""] <- substr(gm_probe[gm$AlleleB_ProbeSeq != ""], 1, 49)
gm_probe_rev <- revcomp(gm_probe)

# subset the before and after sequences to no longer than the probe
nc <- nchar(gm_before_snp)
start <- nc - nchar(gm_probe)+1
start[start < 1] <- 1
gm_before_sub <- gm_before_snp
gm_before_sub[nc>0] <- substr(gm_before_snp[nc>0], start[nc>0], nc[nc>0])

nc <- nchar(gm_after_snp)
end <- nchar(gm_probe)
end[end > nc] <- nc[end > nc]
gm_after_sub <- gm_after_snp
gm_after_sub[nc>0] <- substr(gm_after_snp[nc>0], 1, end[nc>0])

# There were 6 cases where probe doesn't match sequence, but due to lower case letters
# so, convert to upper case
gm_before_sub <- toupper(gm_before_sub)
gm_after_sub <- toupper(gm_after_sub)

stopifnot( all(gm_after_sub == gm_probe_rev | gm_before_sub == gm_probe) )

# also check the strand
stopifnot( all((gm_after_sub == gm_probe_rev & gm$IlmnStrand != gm$SourceStrand) |
               (gm_before_sub == gm_probe & gm$IlmnStrand == gm$SourceStrand)) )

# Now the MegaMUGA
# grab the sequence before and after the SNP
mm_source_spl <- strsplit(mm$SourceSeq, "\\[[ACGT]/[ACGT]\\]")
mm_before_snp <- sapply(mm_source_spl, "[", 1) # a bunch of these have length 0
mm_after_snp <- sapply(mm_source_spl, "[", 2)
mm_after_snp[is.na(mm_after_snp)] <- "" # turn NAs into emptys

# for the SNPs that have both A and B probe sequences, drop the last character from the A sequence
mm_probe <- mm$AlleleA_ProbeSeq
mm_probe[mm$AlleleB_ProbeSeq != ""] <- substr(mm_probe[mm$AlleleB_ProbeSeq != ""], 1, 49)
mm_probe_rev <- revcomp(mm_probe)

# subset the before and after sequences to no longer than the probe
nc <- nchar(mm_before_snp)
start <- nc - nchar(mm_probe)+1
start[start < 1] <- 1
mm_before_sub <- mm_before_snp
mm_before_sub[nc>0] <- substr(mm_before_snp[nc>0], start[nc>0], nc[nc>0])

nc <- nchar(mm_after_snp)
end <- nchar(mm_probe)
end[end > nc] <- nc[end > nc]
mm_after_sub <- mm_after_snp
mm_after_sub[nc>0] <- substr(mm_after_snp[nc>0], 1, end[nc>0])

# There were 9 cases where probe doesn't match sequence, but due to lower case letters
# so, convert to upper case
mm_before_sub <- toupper(mm_before_sub)
mm_after_sub <- toupper(mm_after_sub)

stopifnot( all(mm_after_sub == mm_probe_rev | mm_before_sub == mm_probe) )

# also check the strand
stopifnot( all((mm_after_sub == mm_probe_rev & mm$IlmnStrand != mm$SourceStrand) |
               (mm_before_sub == mm_probe & mm$IlmnStrand == mm$SourceStrand)) )
```

Having gotten here, the probes match the source sequences exactly. SNP
is just after the sequence when `IlmnStrand` is the same as
`SourceStrand`, and is just before the reverse-complement of the
sequence when the strands are different.

## Do the sequences match the UNC files?

We should do a quick check of whether the probe sequences match those in the UNC files.

```{r load_unc_files}
load("../UNC/snps.gigamuga.Rdata")
gm_unc <- snps
load("../UNC/snps.megamuga.Rdata")
mm_unc <- snps
rm(snps)
```

The UNC files for the GigaMUGA and MegaMUGA arrays have
`r add_commas(nrow(gm_unc))` and `r add_commas(nrow(mm_unc))`
rows, respectively. So the MegaMUGA file has the same number of rows
as the file from GeneSeek, but the GigaMUGA file seems to be missing
`r nrow(gm) - nrow(gm_unc)` rows. The missing markers are all have
names that start `CTRL` or `CTRL2`. So control probes, I suppose.

```{r check_same_markers}
stopifnot( all(gm_unc$marker %in% gm$Name) )
stopifnot( all(sort(mm_unc$marker) == sort(mm$Name)) )
```

Otherwise, all of the marker names are the same.

Let's look to see if the sequences are the same.

```{r compare_sequences_geneseek_to_unc}
# sequences in MM files match?
stopifnot( all(mm_unc[mm$Name, "seq.A"] ==  mm$AlleleA_ProbeSeq) )

# subset GeneSeek GM file to those in the UNC file
gm_sub <- gm[gm$Name %in% gm_unc$marker,]
# sequences in GM files match?
stopifnot( all(gm_unc[gm_sub$Name, "seq.A"] ==  gm_sub$AlleleA_ProbeSeq) )
```

Yes, the sequences all match exactly.

I should also compare the chromosome assignments between the GeneSeek
and UNC files.

For the MegaMUGA array, they are the same except some of Y in GeneSeek
are P in UNC, and most of the P and M in UNC are 0 in GeneSeek.

Here's a table just of the markers where they differ.
The rows are the UNC assignments; the columns are the GeneSeek ones.

```{r compare_chr_mm}
chr_lev <- c(0:19,"X","Y","P","M")
mm_unc_chr <- factor(sub("chr", "", mm_unc[mm$Name, "chr"]), chr_lev)
mm_gs_chr <- factor(mm$Chr, chr_lev)
mm_chr_mismatch <- mm_unc_chr != mm_gs_chr
table(mm_unc_chr[mm_chr_mismatch], mm_gs_chr[mm_chr_mismatch])
```

For GM, on the other hand, there are a whole bunch of markers with
disagreement. Also note that the UNC file for the GigaMUGA also has a
bunch of markers with `NA` in the chromosome column. I think they
should be "P", so I'll call them that for now.

```{r compare_chr_gm}
gm_unc_chr <- factor(sub("chr", "", gm_unc[gm_sub$Name, "chr"]), chr_lev)
gm_unc_chr[is.na(gm_unc_chr)] <- "P"
gm_gs_chr <- factor(gm_sub$Chr, chr_lev)
gm_chr_mismatch <- gm_unc_chr != gm_gs_chr
table(gm_unc_chr[gm_chr_mismatch], gm_gs_chr[gm_chr_mismatch])
```


## BLASTn results



## Session info

Here are the versions of R and R packages that I am using.

```{r session_info}
devtools::session_info()
```
