---
title:  "Study SNP sequences on MegaMUGA and GigaMUGA arrays"
author: "Karl Broman"
date:   "`r Sys.Date()`"
output:
    html_document:
        code_folding: hide
        toc: true
        toc_float: true
        toc_depth: 3
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.width=11, fig.height=6.5,
                      message=FALSE, warning=FALSE)
options(width=110)
```


## Preliminaries

I'll first load some R packages.

```{r load_packages}
library(data.table)
library(broman)
```

I'll now load the GeneSeek annotation files for the MegaMUGA and
GigaMUGA arrays. There are a few rows above the header, and then some control
rows at the bottom, that need to be clipped out.

```{r load_geneseek_files}
gm <- data.table::fread("../GeneSeek/gigamuga_geneseek.csv",
                        skip=7, data.table=FALSE)
wh <- which(gm[,1]=="[Controls]")
gm <- gm[1:(wh-1),]
rownames(gm) <- gm$Name

mm <- data.table::fread("../GeneSeek/megamuga_geneseek.csv",
                        skip=7, data.table=FALSE)
wh <- which(mm[,1]=="[Controls]")
mm <- mm[1:(wh-1),]
rownames(mm) <- mm$Name
```

I'll also load the file with the names that are in common between the
two arrays.

```{r load_common_markers}
common <- data.table::fread("../GeneSeek/common_markers.csv", data.table=FALSE)
common_markers <- common[-1,1]
```

## Numbers of markers

The GigaMUGA array contains `r add_commas(nrow(gm))` markers, which
the MegaMUGA has `r add_commas(nrow(mm))` markers.
The marker names in each file
`r ifelse(length(unique(mm$Name)) == nrow(mm) && length(unique(gm$Name)) == nrow(gm), "are", "are not")`
all distinct.
There are `r add_commas(sum(mm$Name %in% gm$Name))` markers in common between
the two arrays.

The Excel file that I received from GeneSeek has three worksheets;
the second and third worksheets contain annotation information, including probe
sequences, for the GigaMUGA and MegaMUGA files, respectively. The first worksheet
indicates which markers are in common between the GigaMUGA and MegaMUGA arrays,
but it's a bit odd. It has counts of markers on each array, which
`r ifelse(common[1,4] == nrow(gm) && common[2,4] == nrow(mm), "are", "are not")`
correct, but there's a count that says "`On Both`",
but it's `r add_commas(common[3,4])`, which is different that what you get in a
direct check of the marker names. Also it has a list of names,
presumably the ones that are in common, and it does have
`r add_commas(nrow(common)-1)` markers, but only there are only
`r add_commas(length(unique(common_markers)))` distinct names, and only
`r add_commas(sum(common_markers %in% gm$Name))` are present in the GigaMUGA
array (`r ifelse(sum(common_markers %in% gm$Name) == sum(common_markers %in% mm$Name), "The same", "A different")`,
number of markers are in the MegaMUGA array.)
The other `r sum(common_markers %nin% gm$Name)` markers contain all
the duplicates and look like control stuff and not real marker names:
`r vec2string(unique(common_markers %wnin% gm$Name))`.
It seems like those came from the `[Control]` section of the other two
worksheets.

The key summaries here:

- GigaMUGA has `r add_commas(nrow(gm))` SNPs
- MegaMUGA has `r add_commas(nrow(mm))` SNPs
- There are `r add_commas(sum(mm$Name %in% gm$Name))` in common


## Available sequences

The GeneSeek annotation files contain four sets of sequences.
`AlleleA_ProbeSeq`, `AlleleB_ProbeSeq`, `SourceSeq`, and `TopGenomicSeq`.

The `AlleleA_ProbeSeq` sequences are
`r ifelse(all(nchar(c(gm$AlleleA_ProbeSeq, mm$AlleleA_ProbeSeq))==50), "all", "*not* all")`
length 50.

The `AlleleB_ProbeSeq` sequences are mostly missing. In the GigaMUGA file there
are `r sum(gm$AlleleB_ProbeSeq != "")` that are not missing; in the MegaMUGA file there
are `r sum(mm$AlleleB_ProbeSeq != "")` that are not missing.

The `SourceSeq` and `TopGenomicSeq` sequences are the same
information. They're identical when the `IlmnStrand` column is `TOP`.
The latter is the reverse complement of the former when the
`IlmnStrand` column is `BOT`.

The `SourceSeq` all seem to contain text like `[A/C]` that indicates
the SNP. This is absent from the `AlleleA_ProbeSeq` which are all
strictly ACGT characters.

## Duplicate probe sequences

There probe sequences are not all distinct. On the GigaMUGA array,
there are `r add_commas(sum(table(gm$AlleleA_ProbeSeq)==2))` that
appear twice, `r sum(table(gm$AlleleA_ProbeSeq)==3)` that appear three
times, `r sum(table(gm$AlleleA_ProbeSeq)==4)` that appear four times.

On the MegaMUGA array,
there are `r add_commas(sum(table(mm$AlleleA_ProbeSeq)==2))` that
appear twice, `r sum(table(mm$AlleleA_ProbeSeq)==3)` that appear three
times, `r sum(table(mm$AlleleA_ProbeSeq)==4)` that appears four times.

## Probe sequences for common markers

Of the `r add_commas(length(common <- mm$Name %win% gm$Name))` markers
that are in common between the two arrays, there are
`r sum(mm[common,"AlleleA_ProbeSeq"] != gm[common, "AlleleA_ProbeSeq"])`
markers that have different probe sequences.

- All cases where one is top strand and other is bottom strand?
- And so one takes sourceseq to left of snp and other to right of snp?
  (because in all cases the snp allele is at end or just after the
  probe sequence?)

## AlleleB probe sequences

For the small number of cases where there's an `AlleleB_ProbeSeq`, in
addition to `AlleleA_ProbeSeq`, I think the last base on the two is
the SNP. They should match at the other 49 bases. Here's a quick check
of that:

```{r check_alleleb_probes}
# cases with AlleleB probe sequence
wh_gm <- which(gm$AlleleB_ProbeSeq != "")
# grab the two sequences
probeA <- gm$AlleleA_ProbeSeq[wh_gm]
probeB <- gm$AlleleB_ProbeSeq[wh_gm]
# first 49 characters are the same
stopifnot( all(substr(probeA, 1, 49) == substr(probeB, 1, 49)) )
# grab the last base from each, which should be the alleles
alleleA <- substr(probeA, 50, 50)
alleleB <- substr(probeB, 50, 50)
# the alleles are different
stopifnot( all( alleleA != alleleB ) )
# the alleles match the SNP column
stopifnot( all( gm$SNP[wh_gm] == paste0("[", alleleA, "/", alleleB, "]") ) )

# now all that for the MegaMUGA array...
# cases with AlleleB probe sequence
wh_mm <- which(mm$AlleleB_ProbeSeq != "")
# grab the two sequences
probeA <- mm$AlleleA_ProbeSeq[wh_mm]
probeB <- mm$AlleleB_ProbeSeq[wh_mm]
# first 49 characters are the same
stopifnot( all(substr(probeA, 1, 49) == substr(probeB, 1, 49)) )
# grab the last base from each, which should be the alleles
alleleA <- substr(probeA, 50, 50)
alleleB <- substr(probeB, 50, 50)
# the alleles are different
stopifnot( all( alleleA != alleleB ) )
# the alleles match the SNP column
stopifnot( all( mm$SNP[wh_mm] == paste0("[", alleleA, "/", alleleB, "]") ) )
```

The code aboved checked that for the cases where there's an AlleleB
probe sequence, the first 49 bases are the same, between the two
alleles, while the 50th base is the SNP

For the remaining probes, I believe that the SNPs is just after the
probe sequence. We'll check that by comparing it to the source
sequences, next.

## Compare probes to source sequences

I'm presuming that the probe sequences match the part of the source
sequence from just before the `[A/T]` snp part.

I'll first split the source sequences into the bits before and after
the snp. I'll also compare the SNP part of the source sequence to the
SNP column. Then I'll compare the parts of the source sequence to
AlleleA probes.

```{r compare_probe_and_source}
source("revcomp.R") # reverse complement

# grab SNP info from SourceSeq, with alleles in both orders
gm_source_snp <- t(sapply(strsplit(gm$SourceSeq, "/"), function(a) {
    n <- nchar(a[1])
    alle1 <- substr(a[1], n, n)
    alle2 <- substr(a[2], 1, 1)
    c(paste0("[", alle1, "/", alle2, "]"), paste0("[", revcomp(alle1), "/", revcomp(alle2), "]")) }))

# SNP column is the same as the source sequence, or is the complement of that
stopifnot( all(gm$SNP == gm_source_snp[,1] | gm$SNP == gm_source_snp[,2]) )

## repeat for MegaMUGA
# grab SNP info from SourceSeq, with alleles in both orders
mm_source_snp <- t(sapply(strsplit(mm$SourceSeq, "/"), function(a) {
    n <- nchar(a[1])
    alle1 <- substr(a[1], n, n)
    alle2 <- substr(a[2], 1, 1)
    c(paste0("[", alle1, "/", alle2, "]"), paste0("[", revcomp(alle1), "/", revcomp(alle2), "]")) }))

# SNP column is the same as the source sequence, or is the complement of that
#    ...but this has no relationship to TOP/BOT in the IlmnStrand column
#    ...need to consider both IlmnStrand and SourceStrand columns
stopifnot( all(mm$SNP == mm_source_snp[,1] | mm$SNP == mm_source_snp[,2]) )

# when SourceStrand=="TOP", the SourceSeq and TopGenomicSeq match
# when SourceStrand=="BOT", the TopGenomicSeq is the reverse complement
#    ...a few cases of lower case letters in the latter
#    ...and it's tricky handling the SNP alleles, b/c they shouldn't be swapped

# grab the sequence before and after the SNP
gm_source_spl <- strsplit(gm$SourceSeq, "\\[[ACGT]/[ACGT]\\]")
gm_before_snp <- sapply(gm_source_spl, "[", 1) # a bunch of these have length 0
gm_after_snp <- sapply(gm_source_spl, "[", 2) # a bunch of these are NA



```




## extra stuff

- is it obvious which base is the SNP? And how they match the SourceSeq?

- seems like snp is just *after* the probe sequence
  sometimes probe B has sequence; seems like then the probe sequence includes the snp
  probe sequences are always 50 bp

I guess check:
- when allele B probe sequence present, the sequences are identical except for last base,
  (and that last base is the SNP allele)
